semihost/machine/mips/mips_semihost.h

```diff
@@ -0,0 +1,394 @@
+    switch (n) {
+    case TARGET_ERRNO_EACCESS:
+        errno = EACCES;
+        break;
+    case TARGET_ERRNO_EAGAIN:
+        errno = EAGAIN;
+        break;
+    case TARGET_ERRNO_EBADF:
+        errno = EBADF;
+        break;
+    case TARGET_ERRNO_EBADMSG:
+        errno = EBADMSG;
+        break;
+    case TARGET_ERRNO_EBUSY:
+        errno = EBUSY;
+        break;
+    case TARGET_ERRNO_ECONNRESET:
+        errno = ECONNRESET;
+        break;
+    case TARGET_ERRNO_EEXIST:
+        errno = EEXIST;
+        break;
+    case TARGET_ERRNO_EFBIG:
+        errno = EFBIG;
+        break;
+    case TARGET_ERRNO_EINTR:
+        errno = EINTR;
+        break;
+    case TARGET_ERRNO_EINVAL:
+        errno = EINVAL;
+        break;
+    case TARGET_ERRNO_EIO:
+        errno = EIO;
+        break;
+    case TARGET_ERRNO_EISDIR:
+        errno = EISDIR;
+        break;
+    case TARGET_ERRNO_ELOOP:
+        errno = ELOOP;
+        break;
+    case TARGET_ERRNO_EMFILE:
+        errno = EMFILE;
+        break;
+    case TARGET_ERRNO_EMLINK:
+        errno = EMLINK;
+        break;
+    case TARGET_ERRNO_ENAMETOOLONG:
+        errno = ENAMETOOLONG;
+        break;
+    case TARGET_ERRNO_ENETDOWN:
+        errno = ENETDOWN;
+        break;
+    case TARGET_ERRNO_ENETUNREACH:
+        errno = ENETUNREACH;
+        break;
+    case TARGET_ERRNO_ENFILE:
+        errno = ENFILE;
+        break;
+    case TARGET_ERRNO_ENOBUFS:
+        errno = ENOBUFS;
+        break;
+    case TARGET_ERRNO_ENOENT:
+        errno = ENOENT;
+        break;
+    case TARGET_ERRNO_ENOMEM:
+        errno = ENOMEM;
+        break;
+    case TARGET_ERRNO_ENOSPC:
+        errno = ENOSPC;
+        break;
+    case TARGET_ERRNO_ENOSR:
+        errno = ENOSR;
+        break;
+    case TARGET_ERRNO_ENOTCONN:
+        errno = ENOTCONN;
+        break;
+    case TARGET_ERRNO_ENOTDIR:
+        errno = ENOTDIR;
+        break;
+    case TARGET_ERRNO_ENXIO:
+        errno = ENXIO;
+        break;
+    case TARGET_ERRNO_EOVERFLOW:
+        errno = EOVERFLOW;
+        break;
+    case TARGET_ERRNO_EPERM:
+        errno = EPERM;
+        break;
+    case TARGET_ERRNO_EPIPE:
+        errno = EPIPE;
+        break;
+    case TARGET_ERRNO_ERANGE:
+        errno = ERANGE;
+        break;
+    case TARGET_ERRNO_EROFS:
+        errno = EROFS;
+        break;
+    case TARGET_ERRNO_ESPIPE:
+        errno = ESPIPE;
+        break;
+    case TARGET_ERRNO_ETIMEDOUT:
+        errno = ETIMEDOUT;
+        break;
+    case TARGET_ERRNO_ETXTBSY:
+        errno = ETXTBSY;
+        break;
+    case TARGET_ERRNO_EXDEV:
+        errno = EXDEV;
+        break;
+    default:
+        errno = EINVAL;
+        break;
+    }
+}
+
+/* Assembler consider sdbbp R2+ */
+#if __mips_isa_rev < 2
+#define MIPS_ISA_LEVEL "mips32r2"
+#else
+#define MIPS_ISA_LEVEL "arch=default"
+#endif
+
+#define UHI_CALL_CODE		"1"
+#define UHI_CALL_CODE_RAW	1
+
+static inline long mips_semihost0(long n)
+{
+    register long r2 __asm__("$2");
```

Comment:
I wonder if it would be better to just call mips_semihost1(n, 0); here and so on to just have one inline ASM block. Setting a few extra argument registers to zero shouldn't really matter for performance. What do you think?

Reply:
I'm just following the convention of other architectures :-) Maybe makes sense.

Reply:
So my preference is to leave it as is but I'm also open to change it if it's a blocker :-)

Reply:
I'm easy with whatever you prefer; this isn't exactly a critical area of the library for size or speed.